### ISSUE TITLE
Support for volume per replica similar to StatefulSets

### ISSUE THREAD

**Issue by ZachTB123** [microsoft/azure-container-apps#357]
Azure Container Apps already supports mounting volumes per [this](https://docs.microsoft.com/en-us/azure/container-apps/storage-mounts?pivots=aca-cli) document. As far as I know, the volume is shared by all replicas in the container app, similar to a K8s Deployment. I need each replica to have its own volume, similar to how [StatefulSets work](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-storage). I'm specifically trying to run an [OpenTelemetry collector](https://github.com/open-telemetry/opentelemetry-collector) in Azure Container Apps using a [persistent queue](https://github.com/open-telemetry/opentelemetry-collector/tree/v0.58.0/exporter/exporterhelper#persistent-queue). My plan was to mount an Azure File share to the container app and use that as my [filestorage extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.58.0/extension/storage/filestorage), which is needed for persistent queuing. It works without issue if I only have one replica in my container app, but as soon as I add more replicas, it does not work. Per [this](https://github.com/open-telemetry/opentelemetry-collector/issues/5894#issuecomment-1212084389) issue, the collector instance expects it is the sole consumer of the queue, so multiple replicas will not be able to consume from it. The suggestion was to run as a StatefulSet or handle it by the mounting directory name. If StatefulSets could not be implemented in Azure Container Apps, I think I could make this work my creating N number of volumes (where N is the max number of replicas I will have, i.e. 30) and mounting them at a directory that includes this number, i.e. `/var/lib/otelcol/1/file_storage`, `/var/lib/otelcol/2/file_storage`, `/var/lib/otelcol/3/file_storage`, etc. Then in my [configuration](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.58.0/extension/storage/filestorage#example) I would set `extensions.file_storage.directory` to `/var/lib/otelcol/$REPLICA_NUMBER/file_storage` to tie the replica to a specific directory. `REPLICA_NUMBER` is the replica number, i.e. with three replicas the first replica's value of `REPLICA_NUMBER` is `1`, the second replica is `2`, and the third replica is `3`. Ultimately I just need N unique, consistent values, and well-known values and I thought replica number would be easy. This would cause the first replica to use the `/var/lib/otelcol/1/file_storage` directory, the second replica to use the `/var/lib/otelcol/2/file_storage`, etc. I would need an environment variable like this made available by the service.

We can definitely run this in AKS to accomplish what we need, but we have certain workloads where we don't want to run AKS and would prefer to use a serverless container service like Azure Container Apps.

In summary: Add support for a volume per replica like StatefulSets or have the service expose an environment variable for the replica number (if I am scaled out to three replicas, each replica would have access to an environment variable named `REPLICA_NUMBER` and the value would be `1` in the first replica, `2` in the second replica, and `3` in the third replica).

**Comment by torosent** [comment]
Thank you for the suggestion @ZachTB123 . We will discuss it internally. Functionality similar to statefulsets is not planned at the moment.

**Comment by csteeg** [comment]
In the meantime I've created a startup script to do this:
```bash
#!/bin/bash

########################
# Retries a command a given number of times
# Arguments:
#   $1 - cmd (as a string)
#   $2 - max retries. Default: 12
#   $3 - sleep between retries (in seconds). Default: 5
# Returns:
#   Boolean
#########################
retry_while() {
    local cmd="${1:?cmd is missing}"
    local retries="${2:-45}"
    local sleep_time="${3:-1}"
    local return_value=1

    read -r -a command <<<"$cmd"
    for ((i = 1; i <= retries; i += 1)); do
        "${command[@]}" && return_value=0 && break
        sleep "$sleep_time"
    done
    return $return_value
}

get_replica_index() {
  local access_token_response=$(curl -sS -H "X-IDENTITY-HEADER: $IDENTITY_HEADER" "${IDENTITY_ENDPOINT}?resource=https://management.azure.com&api-version=2019-08-01")
  if [ "$access_token_response" == *"error"* ]; then
    echo "Error getting access_token: ${access_token_response}"
    return 1
  fi

  JQ=/tmp/jq
  if [ ! -f $JQ ]; then
    curl https://stedolan.github.io/jq/download/linux64/jq > $JQ && chmod +x $JQ
  fi

  local access_token=$(echo $access_token_response | $JQ -r '.access_token')
  if [ ! $access_token ]; then
    echo "No access token! ($access_token)"
    return 1
  fi

  #if no containerapp_id has been set, we assume we can use the xms_mirid(https://github.com/MicrosoftDocs/azure-docs/issues/32131) from the access_token
  if [ -z "$CONTAINERAPP_ID" ]; then
    IFS='.' read -ra JWT_PARTS <<< "$access_token"
    local decoded_payload=$(echo "${JWT_PARTS[1]}" | base64 -d)
    local CONTAINERAPP_ID=$(echo $decoded_payload | $JQ -r '.xms_mirid')
    export CONTAINERAPP_ID=$CONTAINERAPP_ID
  fi

  local get_revision_endpoint="https://management.azure.com$CONTAINERAPP_ID/revisions/$CONTAINER_APP_REVISION?api-version=2022-03-01"
  local revision_response=$(curl -s -H "Authorization: Bearer $access_token" $get_revision_endpoint)
  local replica_count=$(echo "$revision_response" | $JQ -r ".properties.replicas")
  
  if [ "$replica_count" -gt 1]; then
    #"Waiting a second to have all replica instances started and the replica's api returning all instances"
    sleep 1
    local get_replicas_endpoint="https://management.azure.com$CONTAINERAPP_ID/revisions/$CONTAINER_APP_REVISION/replicas?api-version=2022-03-01"
    local replicas_response=$(curl -s -H "Authorization: Bearer $access_token" $get_replicas_endpoint)
    local index=$(echo "$replicas_response" | $JQ -r ".value | sort_by(.name) | sort_by(.properties.createdTime) | map(.name == \"${HOSTNAME}\") | index(true)")
    if [ "$index" == "null" ]; then
      echo "Failed to find $HOSTNAME in replicas:\n $replicas_response"
      return 1
    fi
    export REPLICA_INSTANCE_NO=$((index+1))
    echo "Found replica instance number $REPLICA_INSTANCE_NO"
  else
    echo "Only 1 replica active, so we are number 1 always"
    export REPLICA_INSTANCE_NO=1
  fi
  export REPLICA_INSTANCE_COUNT=$replica_count
  export ACCESS_TOKEN=access_token
}

if ! retry_while "get_replica_index"; then
  echo "Failed to get a replica instance number!"
  exit 1
fi
````

It uses the azure management api to determine the instance-number, so you have to enable managed identity and give access to the instance itself:
I use a module for that, that gets executed after creating the containerapp:
```bicep
param resourceName string
param principalId string
param roleDefinitionId string = resourceId('Microsoft.Authorization/roleDefinitions', 'ad2dd5fb-cd4b-4fd4-a9b6-4fed3630980b') //ContainerApp reader https://www.azadvertizer.net/azrolesadvertizer/ad2dd5fb-cd4b-4fd4-a9b6-4fed3630980b.html

//https://github.com/Azure/bicep/issues/2246 -> we have to strong type to keyvault here instead of generic add-role module
//a switch for vault/storage is also not working: https://github.com/microsoft/azure-container-apps/issues/391#issuecomment-1401513562
resource resource 'Microsoft.App/containerApps@2022-10-01' existing = {
  name: resourceName
}

resource roleAssignment 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = {
  name: guid(resource.id, principalId, roleDefinitionId)
  scope: resource
  properties: {
    roleDefinitionId: roleDefinitionId
    principalId: principalId
    principalType: 'ServicePrincipal'
  }
}

output roleAssignmentId string = roleAssignment.id
```

**Comment by simonjj** [comment]
As @torosent has indicated we're not really looking to support this functionality. We now offer a OTel collector built in. Alternatively we suggest using multiple apps or multiple containers per app possibly. Also, maybe vertical scaling could help. Closing for now. If none of these work for you, please reopen/open a new issue.
